python/ray/data/dataset.py

```diff
@@ -2684,37 +2676,26 @@ def write_datasource(
                 soft=False,
             )
 
-        blocks, metadata = zip(*self._plan.execute().get_blocks_with_metadata())
+        def transform(blocks: Iterable[Block], task_idx, fn):
+            try:
+                datasource.do_write(blocks, task_idx, **write_args)
+                datasource.on_write_complete([])
+            except Exception as e:
+                datasource.on_write_failed([], e)
+                raise
```

Comment:
This isn't the same semantics right? on_write_complete/failed should be called on the driver only, not once per block.

Reply:
You'd need these to implement commit semantics / rollback changes on failure, for example. They are part of the DataSource public API, so we shouldn't change them without a separate API discussion.

Reply:
I feel it's too rush for 2.3 release if we need to change `Datasource` API at this moment, and agree that separate discussion needs to happen if we change `Datasource` API here. How about we taking a non-intrusive approach to not change `Datasource` API at all?

How about we are doing this:

```py
if hasattr(datasource, "_direct_write"):
  def transform(blocks: Iterable[Block], ctx) -> Iterator[Block]:
    yield from datasource._direct_write(blocks, ctx, **write_args)
  self._plan = self._plan.with_stage(transform, ...)
  write_results = self._plan.execute().get_blocks()
else:
  # Old existing code path, or we implement a AllToAllStage, that should give us Block and BlockMetadata
  write_results = ...

try:
  progress.block_until_complete(write_results)
  datasource.on_write_complete(ray.get(write_results))
except Exception as e:
  datasource.on_write_failed(write_results, e)
```

Reply:
I think there is no issue blocked by not having write operator/fusion at the moment, so I'd fix it properly instead of having a half baked feature (2.3 is not the ultimate cut to catch anyway).

Reply:
Yeah, I think we don't need this for 2.3.

Properly supporting the original callback semantics should be possible within the PhysicalOperator framework I believe, though you might need a custom subclass of the map operator.

Reply:
I'm thinking we may use the results of the write tasks to indicate if the entire write operation is complete or failed this way:
- The write tasks can have an output anyway: i.e. the `transform_fn` will be mapping `Iterator[Blocks]` to `WriteResult`. If the task succeeds, `WriteResult` is "ok", otherwise it's the error object from the task.
- The `write_xx()` still returns None, but internally it will run the plan with the write as the last operator; the results of executing that plan is a Dataset comprised of `List[WriteResult]`.
- We don't expose this dataset `List[WriteResult]` from `write_xx()`, but internally use it as an indicator to run these two callbacks: if all elements are "ok", `on_write_complete`, otherwise `on_write_failed`.

This seems address the callback semantics?

Reply:
That makes sense, assuming the result isn't literally "ok" but the user-defined write return value.

Reply:
Hmm, it looks we cannot have the `transform_fn` return the error -- it has to raise the error, in order to trigger the failed tasks retry (we caught this in unit test: https://buildkite.com/ray-project/oss-ci-build-pr/builds/11123).

The right place seems to be the PhysicalOperator as you suggested, as they launch and handle tasks directly. We may add `on_operator_failed()` and `on_operator_complete()` to its interface, and plumb the callback functions to the operator similar to how we plumb the `transform_fn`.

Reply:
It's probably ok to raise the error still I think with the proposed strategy, is there a reason why you had to return them?

Reply:
Returning (instead of raising) errors was to make the above approach (the write operator producing a "status Dataset"), but this is not working for retry.

Now as we agree to add interface to PhysicalOperator to handle this, then we will still raise the error, so Ray will know it should retry. I'm going to change it to this way.

Reply:
I think the status Dataset approach still works right? Errors can get raised as usual, so retires will work. In the write function here we can add a try catch around the entire Dataset execution operation, and call the right data source callbacks if the entire thing fails / succeeds.

Reply:
Oh, do you mean it's hard to provide List[ObjectRef[WriteResult]] to the error callback? I think we can probably deprecate this part of the API for the new backend (pass empty list). Afaik it's not really useful and would be best effort in any scenario for a distributed system.

Reply:
Right, we will not know the resulting status of each individual tasks, which I think may be used for error handling (rollback etc.) in the callback function. If we just try catch the entire execution, and just produce an empty list (as the result of executing write operator), then yes it should be able to work.

Reply:
How about we go for that then? It seems the simplest approach, and we can also fix the new API to not have the list of error refs.

Reply:
Sounds good, I'll update the PR.